web/src/message_view.js

```diff
@@ -234,54 +234,51 @@ function handle_post_message_list_change(
-        // specified via `near` operator, which doesn't handle
-        // `then_select_id` or `then_select_offset` to avoid targeting
-        // the wrong message.
-        return false;
-    }
-
+export function try_rendering_locally_for_same_narrow(filter, opts) {
     const current_filter = narrow_state.filter();
     if (!current_filter) {
         return false;
     }
 
-    if (filter.has_operator("near")) {
-        const target_id = Number.parseInt(filter.operands("near")[0], 10);
-        const target_message = message_lists.current?.get(target_id);
-        if (!target_message) {
-            return false;
-        }
+    let target_id;
+    if (opts.then_select_id !== undefined) {
+        target_id = opts.then_select_id;
+    } else if (filter.has_operator("near")) {
+        target_id = Number.parseInt(filter.operands("near")[0], 10);
+    } else {
+        return false;
+    }
 
-        const adjusted_terms = Filter.adjusted_terms_if_moved(filter.terms(), target_message);
-        if (adjusted_terms !== null) {
-            filter = new Filter(adjusted_terms);
-        }
+    const target_message = message_lists.current?.get(target_id);
+    if (!target_message) {
+        return false;
+    }
 
-        // If the difference between the current filter and the new filter
-        // is just a `near` operator, or just the value of a `near` operator,
-        // we can render the new filter without a rerender of the message list
-        // if the target message in the `near` operator is already rendered.
-        const excluded_operators = ["near"];
-        if (!filter.equals(current_filter, excluded_operators)) {
-            return false;
-        }
+    const adjusted_terms = Filter.adjusted_terms_if_moved(filter.terms(), target_message);
+    if (adjusted_terms !== null) {
+        filter = new Filter(adjusted_terms);
+    }
 
-        const currently_selected_id = message_lists.current?.selected_id();
-        if (currently_selected_id !== target_id) {
-            message_lists.current.select_id(target_id, {
-                then_scroll: true,
-            });
-        }
+    // If the difference between the current filter and the new filter
+    // is just a `near` operator, or just the value of a `near` operator,
+    // we can render the new filter without a rerender of the message list
+    // if the target message in the `near` operator is already rendered.
+    const excluded_operators = ["near"];
+    if (!filter.equals(current_filter, excluded_operators)) {
+        return false;
+    }
 
-        message_lists.current.data.filter = filter;
-        update_hash_to_match_filter(filter, "retarget message location");
-        return true;
+    const currently_selected_id = message_lists.current?.selected_id();
+    if (currently_selected_id !== target_id) {
+        message_lists.current.select_id(target_id, {
+            then_scroll: true,
+            target_scroll_offset: opts.then_select_offset,
```

Comment:
Probably we should pull this out of `opts` above, just to manage risk that some future refactor causes this to be potentially be set when we've not decided to use `then_select_id`.

Reply:
Updated